### PR TITLE
Switch to PyTorch 1.6 in Drone CI

### LINKED ISSUE #4484: Update drone to pytorch 1.6

### PR DESCRIPTION
**Pull request by ydcjeff**
## What does this PR do?

Follow up of #3658 

Fixes #4484 

Related: https://app.circleci.com/pipelines/github/pytorch/pytorch-integration-testing/74/workflows/4a7bc96a-c550-4f0e-8c51-eb9d9af6166c/jobs/451/tests

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by ydcjeff** [comment]
1.7 is causing the failed tests.

**Comment by ydcjeff** [comment]
Can someone confirm our EvalModelTemplate doesn't work in cuda.amp?

It uses nll_loss as loss function which can't be cast to float16, so that can cause loss to be NaN
also there is batch_norm layer which is not listed, so that can also cause Nan loss
https://pytorch.org/docs/master/amp.html#op-specific-behavior

https://github.com/pytorch/pytorch/issues/40497

in the run with precision 32, tests passed: http://104.154.220.231/PyTorchLightning/pytorch-lightning/13333
with precision 16, tests failed: http://104.154.220.231/PyTorchLightning/pytorch-lightning/13330

so, how shall we proceed? change the loss, model or temporarily disable the test?

**Comment by SeanNaren** [comment]
Nice find, @ydcjeff what do you think about swapping the test to use the BoringModel instead? This way we get around the apex requirement for nll (I think we should try our best to not introduce APEX dependencies in our code)

File: tests/models/test_horovod.py [hunk 1]
```diff
@@ -139,11 +142,35 @@ def test_horovod_amp(tmpdir):
         gpus=2,
         deterministic=True,
         distributed_backend='horovod',
+        amp_backend='apex',
```
Comment (Borda) [inline tests/models/test_horovod.py:145]: it means that Horovod as we have it now does not work with AMP from PT 1.6, correct?

Comment (ydcjeff) [reply tests/models/test_horovod.py:145]: Yes, it doesn't work with torch.cuda.amp

Comment (SeanNaren) [reply tests/models/test_horovod.py:145]: Can confirm, this is just an issue with Horovod.

**Comment by SeanNaren** [comment]
hey @ydcjeff how about we upgrade to PyTorch 1.7, but ensure that the horovod tests are skipped for 1.7?

**Comment by ydcjeff** [comment]
> hey @ydcjeff how about we upgrade to PyTorch 1.7, but ensure that the horovod tests are skipped for 1.7?

I have tried with 1.7 and have some DDP issues which are somewhat new to me
maybe we should skip that horovod + amp test for now until horovod release a patch?

**Comment by SeanNaren** [comment]
@ydcjeff I agree, I think we should uncomment the precision=16 line (since this makes the test incorrect) and skip this test with a message saying that until horovod patches torch native integration we'll leave this test. 

I looked at the Horovod repo and it seems its only been tested with APEX mixed precision,  so we could potentially make an issue over there to track integrating torch native amp!

File: tests/models/test_horovod.py [hunk 2]
```diff
@@ -147,28 +146,28 @@ def test_horovod_apex(tmpdir):
...
+#         max_epochs=1,
+#         limit_train_batches=0.4,
+#         limit_val_batches=0.2,
+#         gpus=2,
+#         deterministic=True,
+#         distributed_backend='horovod',
+#         amp_backend='native',
+#         precision=16,
+#     )
+#     _run_horovod(trainer_options, on_gpu=True)
```
Comment (SeanNaren) [inline tests/models/test_horovod.py:170]: Why not add an additional pytest skip?
```suggestion
@pytest.mark.skip(reason="Skip till Horovod fixes integration with Native torch.cuda.amp")
@pytest.mark.skipif(platform.system() == "Windows", reason="Horovod is not supported on Windows")
@pytest.mark.skipif(not _nccl_available(), reason="test requires Horovod with NCCL support")
@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
@pytest.mark.skipif(not NATIVE_AMP_AVALAIBLE, reason="test requires torch.cuda.amp")
def test_horovod_amp(tmpdir):
    """Test Horovod with multi-GPU support using native amp."""
    trainer_options = dict(
        default_root_dir=str(tmpdir),
        weights_save_path=str(tmpdir),
        gradient_clip_val=1.0,
        progress_bar_refresh_rate=0,
        max_epochs=1,
        limit_train_batches=0.4,
        limit_val_batches=0.2,
        gpus=2,
        deterministic=True,
        distributed_backend='horovod',
        amp_backend='native',
        precision=16,
    )
    _run_horovod(trainer_options, on_gpu=True)
```

**Review by SeanNaren — CHANGES_REQUESTED** [review]

File: tests/models/test_horovod.py (hunk 2 above)
Comment (ydcjeff) [reply tests/models/test_horovod.py:170]: Ohh yeah! I'm used to comment so...

**Review by SeanNaren — APPROVED** [review]

**Comment by SeanNaren** [comment]
@ydcjeff we should be good to go right? We'll wait for the build to finish to verify. The next step is to get PyTorch 1.7, but we should probably handle this in a separate PR.

**Comment by ydcjeff** [comment]
> @ydcjeff we should be good to go right? We'll wait for the build to finish to verify. The next step is to get PyTorch 1.7, but we should probably handle this in a separate PR.

yep, it is good to go

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
> @ydcjeff we should be good to go right? We'll wait for the build to finish to verify. The next step is to get PyTorch 1.7, but we should probably handle this in a separate PR.

well there is one PR to run PT 1.6 and 1.7 by @ydcjeff

**Comment by ydcjeff** [comment]
GPU tests are failing...

**Comment by s-rog** [comment]
Thought I'd list the current failing tests here to save some time:
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/plugins/test_amp_plugin.py#L93
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/plugins/test_amp_plugin.py#L123
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/trainer/optimization/test_manual_optimization.py#L222

File: tests/plugins/test_amp_plugin.py [hunk 3]
```diff
@@ -108,7 +108,7 @@ def on_after_backward(self):
         limit_test_batches=2,
         limit_val_batches=2,
         amp_backend='native',
-        distributed_backend='ddp_spawn',
+        distributed_backend='ddp',
```
Comment (SeanNaren) [inline tests/plugins/test_amp_plugin.py:111]: I don't think we can do this, this will make the build hang. It has to remain ddp_spawn

Comment (ydcjeff) [reply tests/plugins/test_amp_plugin.py:111]: spawn has pickle issue but will change back to it

Comment (SeanNaren) [reply tests/plugins/test_amp_plugin.py]: I just made a change to the branch, should fix the pickling errors

Comment (ydcjeff) [reply tests/plugins/test_amp_plugin.py]: Okay thank you Sean

**Comment by ydcjeff** [comment]
The failing amp test is `scaler.update()` is never called as `manual_backward()` only does `.scale()`, `.backward()` and `.unscale_()`
but `scaler.update()` should be called after all optimizers have called `.step()` before calling another `.unscale_()`

**Comment by SeanNaren** [comment]
Hey @ydcjeff this bug was introduced in #4441, I'm going to cc @tchaton to have a look.

@tchaton it seems that after moving the scaler to the accelerator, we're getting a failing test:

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/trainer/optimization/test_manual_optimization.py#L222

Could you have a look?

**Comment by tchaton** [comment]
> Hey @ydcjeff this bug was introduced in #4441, I'm going to cc @tchaton to have a look.
> 
> @tchaton it seems that after moving the scaler to the accelerator, we're getting a failing test:
> 
> https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/trainer/optimization/test_manual_optimization.py#L222
> 
> Could you have a look? Or let us know how to fix it? I can confirm that reverting this PR fixes this test

Hey @ydcjeff, sorry for the introduced bug. I am on it :)

Best regards,
Thomas Chaton

File: pytorch_lightning/plugins/native_amp.py [hunk 4]
```diff
@@ -40,7 +42,7 @@ def backward(self, closure_loss, optimizer, opt_idx, *args, **kwargs):
         closure_loss = closure_loss.detach()
 
         # unscale gradient to allow analyze within `on_after_backward`
-        if not self.trainer.train_loop.should_accumulate() and self.trainer.train_loop.automatic_optimization:
+        if not self.trainer.train_loop.should_accumulate() and automatic_optimization:
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/native_amp.py:45]: I don't think this is correct? We scale the loss regardless (line 30) and we have to unscale regardless of `automatic_optimization`. Is this unscaled elsewhere or something when `automatic_optimization=False`?

Comment (SeanNaren) [reply pytorch_lightning/plugins/native_amp.py:45]: cc @tchaton

Comment (SeanNaren) [reply pytorch_lightning/plugins/native_amp.py:45]: After offline conversation it seems like this is necessary, as the current behaviour is wrong. If the user calls `self.manual_backward` twice with the same optimizer, we call `scaler.unscale_(optimizer)` twice which throws an error. We only want to unscale gradients when we do the `step` calculation.

I suggest adding a function to the trainer such as `def manual_unscale_grads(optimizer)` which the user should call to unscale gradients before doing the optimizer step that handles unscaling of grads if using AMP. This kinda breaks the purpose of manual_backward handling scaling though, so open to suggestions.

**Review by tchaton — APPROVED** [review]

File: pytorch_lightning/plugins/native_amp.py (hunk 4 above)
Comment (ydcjeff) [reply pytorch_lightning/plugins/native_amp.py:45]: @SeanNaren I think we are not calling `scaler.update()` after every optimizer called `scaler.step()` or `optimizer.step()`. If `manual_optimization` is False, we have no control over what should be called after `optimizer.step()`, so would it be better provide `self.scaler` to end users?
I think this way makes sense for now b/c if they want to do their own optimization, they should also need to handle their own 16 bit for now. Like - we leave every 16 bit to end users except `with autocast():`
```py
loss = self.scaler.scale(loss)
self.manual_backward(loss)
self.scaler.unscale_(opt_a)
opt_a.step()
opt_a.zero_grad()
```

If we don't want users to handle 16 bit when `automatic_optimization=False`, we should rethink about calling `manual_backward` + `automatic_optimization` + 16 bit
